### ISSUE TITLE
Festure Request: Support H2O-3 as Backend

### ISSUE THREAD

**Issue by jtelleriar** [tidymodels/parsnip#287]
Would it be possible to support as parsnip engine h2o?

https://cran.r-project.org/web/packages/h2o/index.html

**Comment by stevenpawley** [comment]
FYI I have a rough package in development that provides h2o as an engine:

https://github.com/stevenpawley/h2oparsnip

I haven't added multi_predict methods for the tree-based methods yet, but otherwise, I have been using it for my own modelling purposes. One thing is you cannot simply `saveRDS` a parsnip model_fit if it is using h2o as an engine. It probably needs more thought, but I made a quick `write_h2o` and `read_h2o` function in the package that saves the h2o model and inserts a saved model back into a previously fitted `model_fit` or `workflow` object.

**Comment by topepo** [comment]
Wow. This is on our to-do list. I'll take a look. Also paging @ledell to take a look and offer advice (since we have spoken about this in the past). 

There are some broad questions that we haven't worked out yet (like your example of saving the model). Another example, should we return the data as an R object or as an h2o data object?  

Tuning is going to be another issue; we want to avoid a lot of back-and-forth of data between R and h2o. My thinking was that the tune code could send out some sort of model list for the h2o server to execute, then we can bring the results back in the same form as the current tuning functions.

**Comment by stevenpawley** [comment]
Thanks, I had started a package mainly because I wanted to use the deep learning algorithm as an engine, but thanks to your work on parsnip with others were quick to implement. However, these are only the basic model specifications and don't take advantage of what h2o can do.

Regarding R objects vs. h2o data objects, I had been assuming that if someone is using h2o via parsnip, they are wanting to work with R objects and using h2o behind the scenes?

Tuning is definitely less efficient through `tune_grid` if you are using a remote cluster. I had been thinking about tuning that can also involve recipes, but I wasn't considering changes to tune itself. AFAIU, data from each analysis set (potentially juiced) has to be sent to H2O, but then multiple models can be run for different hyperparameters within H2O without resending any data to the cluster. Still data moving around but less than repeatedly sending data for new models created for each hyperparameter. I had been thinking that this might be accommodated by multi_predict? Although these functions do not have access to the training data, the training data can be kept in the cluster and the multi_predict methods for h2o engines could trigger retraining and prediction based on the new hyperparameters, but using the same H2OFrame id. A problem is how to know when it is safe to remove the H2OFrame used for training. Also, tune will see everything other than parameters in recipes as one model.

**Comment by topepo** [comment]
After looking at this for a bit, I think that we'd make a new code path in `tune` for h2o models. It doesn't seem like h2o can simultaneously tune the pre-processors so we could do those computations on our side and then send the data to `h2o.grid()` for the model computations. That involves a lot of minutiae but is doable. 

We might want to refactor `tune` a bit for this. There are a lot of different "code paths" that can get executed right now (depending on what you are tuning, recipe vs formula, and so on). This, and some other things that we are working on, might expand that a lot and we should probably have a more elegant way of extending the computations. 

I'm not sure that we can do `multi_predict()` here. That happens when the `predict()` method parameterizes a tuning parameter. For GBM, we would need to be able to get the predictions over  many values of the number of trees with one call (or, at least without re-fitting the model). 

> Regarding R objects vs. h2o data objects, I had been assuming that if someone is using h2o via parsnip, they are wanting to work with R objects and using h2o behind the scenes?

I think that this is the case but it would help to get more user feedback. We had this conversation with Spark; do we leave the predictions as a spark table or make it a data frame. For Spark, there might be more emphasis for keeping the predictions out of your local memory and that may not be true for h2o (IIRC my previous conversations with Erin).

**Comment by stevenpawley** [comment]
Thanks for your thoughts. FYI I've added multi_predict functions for h2o.gbm, which use `h2o::staged_predict_proba.H2OModel` to get the class probabilities and/or regression value for any tree stage without retraining. It doesn't grapple with the bigger issues but its nicer when working locally.

**Comment by topepo** [comment]
@stevenpawley We'll probably have interns working on broad H2O support to tidymodels this summer. That would include using the functions in tune (and maybe even finetune) in addition to what you have in h2oparsnip. It might make sense to merge all the H2O/tidymodels functionality into the same package. How would that sound?

**Comment by stevenpawley** [comment]
Hi Max, that would be great - happy to have any parts of h2oparsnip brought into the tidymodels infrastructure. My efforts have stalled for a little while, especially because some input/discussion is needed on issues like how/whether the cluster be managed so that it doesn't accumulate a lot of detritus, or whether h2o frames should be supported directly, and the balance between doing as much as possible on the cluster vs. the flexibility/compatibility with the rest of tidymodels.

**Comment by simonpcouch** [comment]
Closing in light of development on [the agua package](https://github.com/tidymodels/agua)! Thanks all for contributions and input. :)
